### DISCUSSION TITLE
Write register + illegal data address

### DISCUSSION THREAD

**Discussion by Mikail3** [eModbus/eModbus#172] [Q&A] [answered]
Hi there, 

i'm new to modbus and github discussions so correct me if I'm wrong.

I was being able to successfully read out the registers using the **TCP03Example.**
Now I have to be able to **write to these registers.** How is this done?
I've read https://emodbus.github.io/ for some information but could not find anything.

Any help is welcome!

thanks in advance.

**Comment by Miq1** [comment, marked as answer]
Hi!
There are two register write function codes in the standard set: 0x06 WRITE_HOLD_REGISTER and 0x10 WRITE_MULT_REGISTERS. It all depends on which function codes your respective server will support and which addresses it will let be written. There is no general rule for that, there even may be dedicated function codes to modify data.
You may want to have a closer look at the server examples - IIRC some are showing an implementation for the function codes above to write into registers.

**Reply by Mikail3** [reply]
Hi!

Thank you for such a quick response! 
Can u exactly point out to which example exactly uses these functions? 
Because I was not being able to find them.

**Reply by Miq1** [reply]
The RTU06example shows the client part for function code 0x06. RTU of course, but the requests do not differ between RTU and TCP.

**Reply by Mikail3** [reply]
Hi there , sorry to bother again. But I have a question, every time I try to write something to the 0x06 it gives me `Error response: 02 - Illegal data address `

```
 err = MB.addRequest((uint32_t)millis(), 3, READ_HOLD_REGISTER, 41254, 2);
  if (err!=SUCCESS) {
    ModbusError e(err);
    Serial.printf("Error creating request: %02X - %s\n", (int)e, (const char *)e);
  }

  err = MB.addRequest((uint32_t)millis(), 3, WRITE_HOLD_REGISTER, 41254, 381);
  if (err!=SUCCESS) {
    ModbusError e(err);
    Serial.printf("Error creating request: %02X - %s\n", (int)e, (const char *)e);
  }
  
```
  
The read out gets successfully done (it's the same register), and I tried writing to it with different software and that was successful too.
Any idea's about what I'm doing wrong?

sincerely,

**Reply by Miq1** [reply]
The error code 02 is responded by the server you are attempting to write to. I would be quite surprised if other clients could write to that address as you wrote. The eModbus library does not care at all about servers' addresses.
